From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Tue, 16 May 2017 17:51:48 -0400
X-TungstenOS-Note: kernel/sysctl.c hunks dropped. 0083 adds the deny_new_usb entry there and
 0084 moves it out again, so together they leave that file unchanged; keeping them
 would conflict with the disabled unprivileged_userns_clone patch (0064).
Subject: [PATCH 083/101] usb: add toggle for disabling newly added USB devices

Based on the public grsecurity patches.

[thibaut.sautereau@ssi.gouv.fr: Adapt to sysctl code refactoring]
Signed-off-by: Thibaut Sautereau <thibaut.sautereau@ssi.gouv.fr>
Signed-off-by: Levente Polyak <levente@leventepolyak.net>
[thibaut.sautereau@ssi.gouv.fr: Adapt to sysctl code refactoring]
Signed-off-by: Nicolas Bouchinet <nicolas.bouchinet@oss.cyber.gouv.fr>
---
 drivers/usb/core/hub.c |  9 +++++++++
 include/linux/usb.h    |  3 +++
 3 files changed, 26 insertions(+)

diff --git a/drivers/usb/core/hub.c b/drivers/usb/core/hub.c
index 068c7be..23e7918 100644
--- a/drivers/usb/core/hub.c
+++ b/drivers/usb/core/hub.c
@@ -5400,6 +5400,9 @@ static int descriptors_changed(struct usb_device *udev,
 	return changed;
 }
 
+/* sysctl */
+int deny_new_usb __read_mostly = 0;
+
 static void hub_port_connect(struct usb_hub *hub, int port1, u16 portstatus,
 		u16 portchange)
 {
@@ -5461,6 +5464,12 @@ static void hub_port_connect(struct usb_hub *hub, int port1, u16 portstatus,
 			goto done;
 		return;
 	}
+
+	if (deny_new_usb) {
+		dev_err(&port_dev->dev, "denied insert of USB device on port %d\n", port1);
+		goto done;
+	}
+
 	if (hub_is_superspeed(hub->hdev))
 		unit_load = 150;
 	else
diff --git a/include/linux/usb.h b/include/linux/usb.h
index 49ab8db..f2a0543 100644
--- a/include/linux/usb.h
+++ b/include/linux/usb.h
@@ -2113,6 +2113,9 @@ extern void usb_led_activity(enum usb_led_event ev);
 static inline void usb_led_activity(enum usb_led_event ev) {}
 #endif
 
+/* sysctl */
+extern int deny_new_usb;
+
 #endif  /* __KERNEL__ */
 
 #endif
